### ISSUE TITLE
$geoIntersects not implemented

### ISSUE THREAD

**Issue by duncaneddy** [mongomock/mongomock#570]
In the 3.18.0 release there is a regression where some valid mongo operators are reported as:
`pymongo.erros.OperationFailure: unknown operator`

See trace:
```
self = <mongomock.filtering._Filterer object at 0x1363d57f0>
search_filter = {'geometry': {'$geoIntersects': {'$geometry': {'coordinates': [-122.4194, 37.7749], 'type': <GeoJSONGeometryType.Point: 'Point'>}}}}
document = {}

    def apply(self, search_filter, document):
        if not isinstance(search_filter, dict):
            raise OperationFailure('the match filter must be an expression in an object')
    
        for key, search in iteritems(search_filter):
            # Top level operators.
            if key == '$comment':
                continue
            if key in LOGICAL_OPERATOR_MAP:
                if not search:
                    raise OperationFailure('BadValue $and/$or/$nor must be a nonempty array')
                if not LOGICAL_OPERATOR_MAP[key](document, search, self.apply):
                    return False
                continue
            if key in _TOP_LEVEL_OPERATORS:
                raise NotImplementedError(
                    'The {} operator is not implemented in mongomock yet'.format(key))
            if key.startswith('$'):
                raise OperationFailure('unknown top level operator: ' + key)
    
            is_match = False
    
            is_checking_negative_match = \
                isinstance(search, dict) and {'$ne', '$nin'} & set(search.keys())
            is_checking_positive_match = \
                not isinstance(search, dict) or (set(search.keys()) - {'$ne', '$nin'})
            has_candidates = False
    
            if search == {'$exists': False} and not iter_key_candidates(key, document):
                continue
    
            if isinstance(search, dict) and '$all' in search:
                if not self._all_op(iter_key_candidates(key, document), search['$all']):
                    return False
                continue
    
            for doc_val in iter_key_candidates(key, document):
                has_candidates |= doc_val is not NOTHING
                is_ops_filter = search and isinstance(search, dict) and \
                    all(key.startswith('$') for key in search.keys())
                if is_ops_filter:
                    if '$options' in search and '$regex' in search:
                        search = _combine_regex_options(search)
                    unknown_operators = set(search) - set(self._operator_map) - {'$not'}
                    if unknown_operators:
>                       raise OperationFailure('unknown operator: ' + list(unknown_operators)[0])
E                       pymongo.errors.OperationFailure: unknown operator: $geoIntersects

../../../.pyenv/versions/3.6.7/lib/python3.6/site-packages/mongomock/filtering.py:100: OperationFailure
```

Version 3.17.0 does not have this issue, and the query operator `$geoIntersects` in indeed a valid operator https://docs.mongodb.com/manual/reference/operator/query/geoIntersects/

I'll put together a fix and submit a pull request when I have a little time to figure this out.

**Comment by pcorpet** [comment]
I'm not sure it's a regression: the `$geoIntersects` operator was never implemented in mongomock, so before 3.18 it gave you a false feeling that it worked but was actually silently ignoring it. We should report this as a `NotImplementedError` but it will still choke on your code unfortunately.

I'll submit a fix to get a better error message. Feel free to send a real fix to implement the feature if you feel like it :-)

**Comment by duncaneddy** [comment]
Ah, definitely not a regression then. Thank you for the clarification. A real fix is definitely something I'd like to implement, just can't make any promises on when. I suspect that fully implementing the geospatial queries won't be trivial. 

Thank you for making and maintaining this—love the package!

**Comment by wessm** [comment]
I have implemented `$geoIntersects` in a fork for my personal projects, but it is based on the 3rd party library [`shapely`](https://github.com/Toblerity/Shapely). If this is acceptable, I would go ahead and try to create my very first pull request.

**Comment by duncaneddy** [comment]
That's awesome to hear! I still haven't had time to get to this myself, so really glad to hear you have an implementation.

If having that dependency is not acceptable in the default installation, another way of handling that would be to implement shapely as an extra (optional) dependency. And have `$geoIntersects` only work if that package is installed (and otherwise error).

The [setuptools documentation](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) has an example of how that could be done.

Celery [uses this approach](https://docs.celeryproject.org/en/stable/getting-started/introduction.html#bundles) for some of their optional features and could be a nice way to handle it. But I'd of course defer to what @pcorpet would prefer.

**Comment by victorct-pronto** [comment]
Hello, guys first of all thank you so much for the library, we highly use MongoDB and many of their `geo` queries, is there any update on this, if not @wessm I would be happy to see the implementation with `shapely`  and create a PR for the library.

**Comment by Seluj78** [comment]
Hi! any news on this ?

Also, I'm wondering why this wasn't implemented in the first place ? I find a lot of operations aren't available :/

**Comment by pcorpet** [comment]
Sorry for not answering earlier. I like the idea of using `shapely` library if it's installed and throw a `NotImplementedError` telling you to install it if not. @wessm @duncaneddy feel free to send a pull request. Make sure you add a proper Travis check with and without the shapely lib.

@Seluj78 mongomock has no claim on having all features of MongoDB. Check out https://github.com/mongomock/mongomock#important-note-about-project-status--development

**Comment by Seluj78** [comment]
> @Seluj78 mongomock has no claim on having all features of MongoDB. Check out [mongomock/mongomock#important-note-about-project-status--development](https://github.com/mongomock/mongomock#important-note-about-project-status--development)

Fair enough ! I was just wondering and I didn't read through the whole reamde, my bad!

**Comment by rezo1134** [comment]
This looks to have died on the vine. @wessm Any chance your shapely implementation could be updated and still added as a PR ?

**Comment by wessm** [comment]
I am sorry, my implementation was part of an old codebase belonging to one of my former employers, I have no access to it anymore. But if I remember correctly, it was a 15 mins job.

Edit: Not sure why I wrote "personal project" 4 years ago...
